File: ios/RoomFinderAI/Services/ListingDraftService.swift
import UIKit

/// Fills in a listing for the host instead of making them write one.
///
/// Two routes, matching what the site already does:
///   • a photo goes to `/api/analyze-property-photo`, which runs Cloudflare's
///     Llama 3.2 Vision model and reads the room itself
///   • no photo falls back to `/api/listings/draft`, which writes from the
///     facts already typed in
///
/// Writing the description is where people abandon posting a room, so this is
/// the difference between a listing existing and not.
@MainActor
final class ListingDraftService {

    struct Draft {
        var title: String?
        var description: String?
        var houseType: String?
        var bedrooms: Int?
        /// The worker's rent estimate. It already computed this from the property
        /// type, the features it spotted and the postcode, and the app was
        /// throwing it away — leaving the host to guess the one number that
        /// decides whether anyone enquires.
        var price: Int?
        var street: String?
        var city: String?
        var postalCode: String?
        /// True when the address came from an IP lookup rather than GPS, so
        /// the host is told to check it rather than trusting a guess.
        var locationIsApproximate = false
        /// Notable things it saw, for the "what it saw" summary.
        var features: [String] = []
    }

    /// Where a photo was taken, read out of its own metadata.
    struct PhotoLocation {
        let latitude: Double
        let longitude: Double
    }

    enum DraftError: LocalizedError {
        case noInput
        case failed(String)

        var errorDescription: String? {
            switch self {
            case .noInput: return "Add a photo, or fill in the city and property type first."
            case .failed(let message): return message
            }
        }
    }

    /// Reads the room from a photo.
    func draft(from image: UIImage, at location: PhotoLocation? = nil) async throws -> Draft {
        // Downscaled before sending: the endpoint takes the image as a JSON
        // array of bytes, which is roughly four times the size of the raw
        // file, so a full-resolution phone photo would be a multi-megabyte
        // request for no extra accuracy.
        guard let data = Self.downscaledJPEG(image, maxDimension: 900) else {
            throw DraftError.failed("That photo could not be read.")
        }

        var request = URLRequest(url: AppConfig.url("api/analyze-property-photo"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        // Vision models are slow; this is the one place a long wait is expected.
        request.timeoutInterval = 120
        // Base64, not a JSON array of 90,000 boxed numbers.
        //
        // `["image": Array(data)]` made JSONSerialization allocate an NSNumber
        // per byte and emit roughly half a megabyte of text for a 90KB photo.
        // On device that is the difference between a moment and minutes — the
        // server answers this endpoint in about 1.4s, so the wait was almost
        // entirely the encoding.
        var payload: [String: Any] = ["imageBase64": data.base64EncodedString()]
        if let location {
            payload["coords"] = ["lat": location.latitude, "lng": location.longitude]
        }
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)

        let (responseData, response) = try await URLSession.shared.data(for: request)

        struct AnalysisResponse: Decodable {
            struct Analysis: Decodable {
                struct Location: Decodable {
                    let street: String?
                    let city: String?
                    let zip: String?
                    /// "gps" when it came from real coordinates, "ip" when it
                    /// was inferred from the network and is city-level at best.
                    let source: String?
                }
                let title: String?
                let description: String?
                let house_type: String?
                let bedrooms: Int?
                let suggestedPrice: Double?
                let features: [String]?
                let location: Location?
            }
            let success: Bool
            let analysis: Analysis?
            let error: String?
            let rejected: Bool?
        }

        guard let http = response as? HTTPURLResponse else {
            throw DraftError.failed("The photo analysis service could not be reached.")
        }

        // A rejected photo comes back 422 with a specific, useful reason —
        // "this is a styled interior rendering, not a photograph of a real
        // rental property". That used to be discarded by a status-code check
        // that reported "the service could not be reached" instead, so a
        // working validator doing exactly the right thing looked like the
        // feature was broken. Read the body before judging on the status.
        if !(200..<300).contains(http.statusCode) {
            if let body = try? JSONDecoder().decode(AnalysisResponse.self, from: responseData),
               let message = body.error, !message.isEmpty {
                throw DraftError.failed(message)
            }
            throw DraftError.failed("The photo analysis service could not be reached.")
        }

        let decoded = try JSONDecoder().decode(AnalysisResponse.self, from: responseData)
        guard decoded.success, let analysis = decoded.analysis else {
            throw DraftError.failed(decoded.error ?? "The photo could not be analysed.")
        }

        return Draft(
            title: analysis.title,
            description: analysis.description,
            houseType: analysis.house_type,
            bedrooms: analysis.bedrooms,
            price: analysis.suggestedPrice.map { Int($0.rounded()) },
            street: analysis.location?.street,
            city: analysis.location?.city,
            postalCode: analysis.location?.zip,
            locationIsApproximate: analysis.location?.source == "ip",
            features: analysis.features ?? []
        )
    }

    /// Writes from the details already entered, when there is no photo yet.
    func draft(city: String, street: String, houseType: String,
               bedrooms: Int, price: String, utilitiesIncluded: Bool,
               notes: String) async throws -> Draft {
        guard !city.trimmingCharacters(in: .whitespaces).isEmpty else {
            throw DraftError.noInput
        }

        var body: [String: Any] = [
            "city": city,
            "houseType": houseType,
            "bedrooms": bedrooms,
            "utilities": utilitiesIncluded ? "included" : "not included"
        ]
        if !street.isEmpty { body["street"] = street }
        if let value = Int(price.filter(\.isNumber)), value > 0 { body["price"] = value }
        if !notes.isEmpty { body["notes"] = notes }

        var request = URLRequest(url: AppConfig.url("api/listings/draft"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 60
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw DraftError.failed("No response from the server.")
        }

        struct DraftResponse: Decodable {
            let success: Bool?
            let title: String?
            let description: String?
            let message: String?
        }
        let decoded = try? JSONDecoder().decode(DraftResponse.self, from: data)

        guard (200..<300).contains(http.statusCode), decoded?.success == true else {
            throw DraftError.failed(decoded?.message ?? "Could not write a draft right now.")
        }

        return Draft(title: decoded?.title, description: decoded?.description)
    }

    /// Longest edge capped, re-encoded as JPEG.
    private static func downscaledJPEG(_ image: UIImage, maxDimension: CGFloat) -> Data? {
        let longest = max(image.size.width, image.size.height)
        guard longest > 0 else { return nil }

        let scale = min(1, maxDimension / longest)
        let size = CGSize(width: image.size.width * scale, height: image.size.height * scale)

        let renderer = UIGraphicsImageRenderer(size: size)
        let resized = renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: size))
        }
        return resized.jpegData(compressionQuality: 0.75)
    }
}

File: RoomFinderAndroid/app/src/main/java/com/roomfinder/android/services/ListingDraftService.java
package com.roomfinder.android.services;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.util.Base64;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;

/**
 * Fills in a listing for the host instead of making them write one.
 *
 * Two routes, the same ones the website and the iOS app use:
 *   - a photo goes to /api/analyze-property-photo, where a vision model reads
 *     the room itself and comes back with a title, a description, the property
 *     type, a bedroom count, a suggested rent and what it noticed
 *   - with no photo, /api/listings/draft writes from the facts already typed
 *
 * Writing the description is where people abandon posting a room, so this is
 * the difference between a listing existing and not.
 */
public class ListingDraftService {
    private static final String TAG = "ListingDraftService";
    private static final String BASE_URL = "https://www.roomfinderai.com/api/";
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    /** Vision models are slow; this is the one place a long wait is expected. */
    private static final int ANALYSIS_TIMEOUT_SECONDS = 120;

    /**
     * A photo is downscaled before sending. Full-resolution phone photos add
     * megabytes to the request without making the model any more accurate.
     */
    private static final int MAX_DIMENSION = 900;

    private final OkHttpClient client;
    private final Handler main = new Handler(Looper.getMainLooper());

    public static class Draft {
        public String title;
        public String description;
        public String houseType;
        public Integer bedrooms;
        public Integer price;
        public String street;
        public String city;
        public String postalCode;
        /** True when the address came from an IP lookup rather than GPS. */
        public boolean locationIsApproximate;
        /** Notable things the model saw, for the "what it saw" summary. */
        public final List<String> features = new ArrayList<>();

        public boolean hasLocation() {
            return (city != null && !city.isEmpty()) || (street != null && !street.isEmpty());
        }
    }

    public interface DraftCallback {
        void onDraft(Draft draft);

        void onError(String message);
    }

    public ListingDraftService() {
        this.client = new OkHttpClient.Builder()
                .connectTimeout(20, TimeUnit.SECONDS)
                .readTimeout(ANALYSIS_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                .writeTimeout(ANALYSIS_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                .build();
    }

    /** Reads the room from a photo. */
    public void draftFromPhoto(Context context, Uri photoUri, DraftCallback callback) {
        String encoded = encodePhoto(context, photoUri);
        if (encoded == null) {
            callback.onError("That photo could not be read.");
            return;
        }

        try {
            JSONObject payload = new JSONObject();
            payload.put("imageBase64", encoded);

            Request request = new Request.Builder()
                    .url(BASE_URL + "analyze-property-photo")
                    .addHeader("Content-Type", "application/json")
                    .post(RequestBody.create(payload.toString(), JSON))
                    .build();

            client.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    Log.e(TAG, "Photo analysis failed", e);
                    post(() -> callback.onError("The photo analysis service could not be reached."));
                }

                @Override
                public void onResponse(Call call, Response response) {
                    try (ResponseBody body = response.body()) {
                        String raw = body == null ? "" : body.string();

                        // A rejected photo comes back 422 with a specific,
                        // useful reason ("this is a styled interior rendering,
                        // not a photograph of a real rental property"). Judging
                        // on the status code alone threw that away and reported
                        // a connection problem instead.
                        if (!response.isSuccessful()) {
                            String message = errorFrom(raw);
                            post(() -> callback.onError(message != null
                                    ? message
                                    : "The photo could not be analysed."));
                            return;
                        }

                        JSONObject json = new JSONObject(raw);
                        if (!json.optBoolean("success", false) || json.isNull("analysis")) {
                            String message = json.optString("error", "");
                            post(() -> callback.onError(message.isEmpty()
                                    ? "The photo could not be analysed."
                                    : message));
                            return;
                        }

                        Draft draft = parseAnalysis(json.getJSONObject("analysis"));
                        post(() -> callback.onDraft(draft));
                    } catch (Exception e) {
                        Log.e(TAG, "Failed to read analysis", e);
                        post(() -> callback.onError("Couldn't read the photo analysis."));
                    }
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "Failed to build analysis request", e);
            callback.onError("Couldn't analyse that photo.");
        }
    }

    /** Writes from the details already entered, when there is no photo yet. */
    public void draftFromDetails(String city, String street, String houseType, int bedrooms,
                                 String price, boolean utilitiesIncluded, String notes,
                                 DraftCallback callback) {
        if (city == null || city.trim().isEmpty()) {
            callback.onError("Add a photo, or fill in the city and property type first.");
            return;
        }

        try {
            JSONObject payload = new JSONObject();
            payload.put("city", city.trim());
            payload.put("street", street == null ? "" : street.trim());
            payload.put("houseType", houseType == null ? "Apartment" : houseType);
            payload.put("bedrooms", bedrooms);
            payload.put("price", price == null ? "" : price.trim());
            payload.put("utilitiesIncluded", utilitiesIncluded);
            payload.put("notes", notes == null ? "" : notes.trim());

            Request request = new Request.Builder()
                    .url(BASE_URL + "listings/draft")
                    .addHeader("Content-Type", "application/json")
                    .post(RequestBody.create(payload.toString(), JSON))
                    .build();

            client.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    Log.e(TAG, "Draft request failed", e);
                    post(() -> callback.onError("Couldn't reach the writer. Check your connection."));
                }

                @Override
                public void onResponse(Call call, Response response) {
                    try (ResponseBody body = response.body()) {
                        String raw = body == null ? "" : body.string();
                        JSONObject json = new JSONObject(raw);
                        if (!response.isSuccessful() || !json.optBoolean("success", false)) {
                            String message = json.optString("error", "");
                            post(() -> callback.onError(message.isEmpty()
                                    ? "Couldn't write the listing."
                                    : message));
                            return;
                        }

                        Draft draft = new Draft();
                        draft.title = optString(json, "title");
                        draft.description = optString(json, "description");
                        post(() -> callback.onDraft(draft));
                    } catch (Exception e) {
                        Log.e(TAG, "Failed to read draft", e);
                        post(() -> callback.onError("Couldn't read the written listing."));
                    }
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "Failed to build draft request", e);
            callback.onError("Couldn't write the listing.");
        }
    }

    private Draft parseAnalysis(JSONObject analysis) {
        Draft draft = new Draft();
        draft.title = optString(analysis, "title");
        draft.description = optString(analysis, "description");
        draft.houseType = optString(analysis, "house_type");

        if (analysis.has("bedrooms") && !analysis.isNull("bedrooms")) {
            draft.bedrooms = analysis.optInt("bedrooms");
        }
        // The worker already computed a rent from the property type, the
        // features it spotted and the postcode. Throwing it away left the host
        // guessing the one number that decides whether anyone enquires.
        if (analysis.has("suggestedPrice") && !analysis.isNull("suggestedPrice")) {
            draft.price = (int) Math.round(analysis.optDouble("suggestedPrice"));
        }

        JSONArray features = analysis.optJSONArray("features");
        if (features != null) {
            for (int i = 0; i < features.length(); i++) {
                String feature = features.optString(i, "").trim();
                if (!feature.isEmpty()) {
                    draft.features.add(feature);
                }
            }
        }

        JSONObject location = analysis.optJSONObject("location");
        if (location != null) {
            draft.street = optString(location, "street");
            draft.city = optString(location, "city");
            draft.postalCode = optString(location, "zip");
            // The vision worker rewrites "source" to "gps" for anything handed
            // to it, so an IP-derived area arrives claiming to be GPS. The
            // "approximate" flag is set by our own server and survives that.
            draft.locationIsApproximate = location.optBoolean("approximate", false)
                    || "ip".equalsIgnoreCase(optString(location, "source"));
        }
        return draft;
    }

    private String encodePhoto(Context context, Uri photoUri) {
        try (InputStream input = context.getContentResolver().openInputStream(photoUri)) {
            if (input == null) {
                return null;
            }
            Bitmap original = BitmapFactory.decodeStream(input);
            if (original == null) {
                return null;
            }
            Bitmap scaled = downscale(original);
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            scaled.compress(Bitmap.CompressFormat.JPEG, 80, out);
            if (scaled != original) {
                scaled.recycle();
            }
            // Base64, not a JSON array of boxed bytes: the array form is about
            // four times the size for no extra accuracy.
            return Base64.encodeToString(out.toByteArray(), Base64.NO_WRAP);
        } catch (Exception e) {
            Log.e(TAG, "Couldn't encode photo", e);
            return null;
        }
    }

    private Bitmap downscale(Bitmap source) {
        int width = source.getWidth();
        int height = source.getHeight();
        int longest = Math.max(width, height);
        if (longest <= MAX_DIMENSION) {
            return source;
        }
        float ratio = (float) MAX_DIMENSION / longest;
        return Bitmap.createScaledBitmap(source,
                Math.round(width * ratio), Math.round(height * ratio), true);
    }

    private String errorFrom(String raw) {
        try {
            JSONObject json = new JSONObject(raw);
            String message = json.optString("error", "");
            return message.isEmpty() ? null : message;
        } catch (Exception e) {
            return null;
        }
    }

    private static String optString(JSONObject json, String key) {
        if (json == null || !json.has(key) || json.isNull(key)) {
            return null;
        }
        String value = json.optString(key, "").trim();
        return value.isEmpty() ? null : value;
    }

    private void post(Runnable action) {
        main.post(action);
    }
}
